cpp/include/cuvs/neighbors/vamana.hpp

```diff
@@ -0,0 +1,278 @@
+  {
+    RAFT_LOG_DEBUG("Copying Vamana graph from host to device");
+
+    if ((graph_.extent(0) != new_graph.extent(0)) || (graph_.extent(1) != new_graph.extent(1))) {
+      // clear existing memory before allocating to prevent OOM errors on large graphs
+      if (graph_.size()) { graph_ = raft::make_device_matrix<IdxT, int64_t>(res, 0, 0); }
+      graph_ =
+        raft::make_device_matrix<IdxT, int64_t>(res, new_graph.extent(0), new_graph.extent(1));
+    }
+    raft::copy(graph_.data_handle(),
+               new_graph.data_handle(),
+               new_graph.size(),
+               raft::resource::get_cuda_stream(res));
+    graph_view_ = graph_.view();
+  }
+
+ private:
+  cuvs::distance::DistanceType metric_;
+  raft::device_matrix<IdxT, int64_t, raft::row_major> graph_;
+  raft::device_matrix_view<const IdxT, int64_t, raft::row_major> graph_view_;
+  std::unique_ptr<neighbors::dataset<int64_t>> dataset_;
+  IdxT medoid_id_;
+};
+/**
+ * @}
+ */
+
+/**
+ * @defgroup vamana_cpp_index_build Vamana index build functions
+ * @{
+ */
+/**
+ * @brief Build the index from the dataset for efficient search.
+ *
+ */
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::device_matrix_view<const float, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<float, uint32_t>;
+
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::host_matrix_view<const float, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<float, uint32_t>;
+
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::device_matrix_view<const int8_t, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<int8_t, uint32_t>;
+
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::host_matrix_view<const int8_t, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<int8_t, uint32_t>;
+
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::device_matrix_view<const uint8_t, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<uint8_t, uint32_t>;
+
+auto build(raft::resources const& handle,
+           const cuvs::neighbors::vamana::index_params& params,
+           raft::host_matrix_view<const uint8_t, int64_t, raft::row_major> dataset)
+  -> cuvs::neighbors::vamana::index<uint8_t, uint32_t>;
+
+/**
+ * @defgroup vamana_cpp_serialize Vamana serialize functions
+ * @{
+ */
+/**
+ * Save the index to file.
+ */
+
+void serialize(raft::resources const& handle,
+               const std::string& file_prefix,
+               const cuvs::neighbors::vamana::index<float, uint32_t>& index,
+               bool include_dataset = true);
```

Comment:
I had considered having the option to write the index out as well, but didn't implement it yet. The CPU DiskANN repo takes separate files for the dataset and index when running search, though it requires them to be named "<index_name>", and "<index_name>.data". A user could easily just rename the dataset, but it would require some knowledge of this, so writing the filename may make it simpler for some users.
Not sure if this is needed for the initial version, though.

Reply:
Took it out for now, can add an option to output the dataset later if needed.